src/instructlab/model/full_train.py

```diff
@@ -0,0 +1,319 @@
+    logger.info(f"Total RAM: {total_ram:.2f} GB")
+    # if RAM is <= 16, we need to use fp16 not fp32. This will yield a worse model but will allow the full pipeline to run
+    accum = 1
+    if total_ram <= 16:
+        # if <= 16GB ram, use gradinent accum and hald precision
+        logger.warning(
+            f"Your system has {total_ram:.2f} GB of RAM. This is below our reccomendation of 32GB for this type of training. Using half precision."
+        )
+        model = model.to(dev).half()  # Convert model to float16
+        accum = 4
+    else:
+        model = model.to(dev)
+
+    # adafactor and gradient checkpointing are memory friendly, we opt to use these in the CPU/MPS loop to fit 7b models.
+    optimizer = Adafactor(
+        model.parameters(), lr=1e-5, scale_parameter=True, relative_step=False
+    )
+    model.gradient_checkpointing_enable()
+
+    model.train()
+    global_step = 1
+
+    # For each epoch do the following:
+    # 1. Incremement Loading Bar
+    # 2. put zeros onto the proper device so we can store some values we need
+    # 3. put each part of the batch onto the device as well using fp6
+    # 4. Forward pass on the model. If using half, we need to do some special handling. If not, then just put the batch onto the device using fp32
+    # 5. Calculate the loss
+    # 6. backward pass on the loss
+    # 7. We are using gradient accumulation to be memory efficient, so if this step is divisible by 4, take a step and zero out the gradients (freeing memory)
+    # 8. clear out some caches if on MPS
+    # 9. Incrememnt global step and loading bar
+    for epoch in range(train_args.num_epochs):
+        dataloader.batch_sampler.set_epoch(epoch)
+        inner_pb = tqdm(range(len(dataloader)), desc=f"Epoch {epoch}")
+        aggregated_values = torch.zeros(3, dtype=torch.float16).to(dev)
+
+        for step, batch in enumerate(dataloader):
+            aggregated_values[0] = batch.pop("num_loss_counted_tokens")
+            aggregated_values[1] = len(batch["input_ids"])
+
+            # Move and cast batch data to device
+            for k in batch:
+                if total_ram < 16:
+                    if k in ["input_ids", "attention_mask"]:
+                        # these two need to be of type long if using .half()
+                        batch[k] = batch[k].to(device=dev, dtype=torch.long)
+                    else:
+                        batch[k] = batch[k].to(device=dev, dtype=torch.float16)
+                else:
+                    batch[k] = batch[k].to(device=dev)
+
+            output = model(**batch, use_cache=False)
+            loss = output.loss
+            aggregated_values[2] = loss.item()
+
+            num_loss_counted_tokens = aggregated_values[0]
+            loss = loss / num_loss_counted_tokens
+
+            loss = loss / accum  # Scale the loss for accumulation steps
+
+            logger.info(
+                f"\nEpoch: {epoch}, Step: {global_step}, Rank: 0, loss = {loss}"
```

Comment:
should this be `step` instead of `global_step`?

Reply:
step is just from the enum of the data, global step is our tracker of how many steps we have gone thru. They may be the same? but our global steps should start at 1 not 0. I can probably combine these.